File: src/main/java/equidiaryDB/EquidiaryDB.java
package equidiaryDB;

import equidiaryDB.config.Config;
import equidiaryDB.config.NullConfig;
import equidiaryDB.database.DataBase;
import equidiaryDB.services.LoginService;
import io.javalin.Javalin;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.flywaydb.core.Flyway;

import java.nio.file.Path;
import java.nio.file.Paths;

public class EquidiaryDB
{
    private static final int EQUIDIARYDB_PORT = 7001;
    private static final Path CONFIG_PROPERTIES_PATH = Paths.get("config/config.properties");
    public static final Logger LOGGER = LogManager.getLogger();
    private static Javalin app;

    public static void start() throws Exception
    {
        Config config = Config.createConfig(CONFIG_PROPERTIES_PATH);

        if (failedToLoadConfig(config))
        {
            LOGGER.fatal("Failed to start EquidiaryDB");
            return;
        }

//        applyDataBaseMigrations(config);

        DataBase.createDatabase(config.getHostDB(),
                config.getPortDB(),
                config.getUserDB(),
                config.getPasswordDB(),
                config.getSchemaDB());

        app = Javalin.create().start(EQUIDIARYDB_PORT);

        createEndPoints();
    }

    private static void applyDataBaseMigrations(Config config)
    {
        String url = "jdbc:mysql://"
                + config.getHostDB()
                + ":"
                + config.getPortDB()
                + "/"
                + config.getSchemaDB()
                + "?serverTimezone=UTC";

        Flyway db_migration = Flyway.configure()
                .locations("db_migration")
                .dataSource(url, config.getUserDB(), config.getPasswordDB())
                .load();

        db_migration.migrate();
    }

    private static boolean failedToLoadConfig(Config config)
    {
        return config == NullConfig.INSTANCE;
    }

    public static void stop()
    {
        app.stop();
    }

    private static void createEndPoints()
    {
        app.get("/", ctx -> ctx.result("Hello World"));
        app.get("/users/:name", ctx -> ctx.result("Hello " + ctx.pathParam("name")));
        app.post("/login", LoginService::login);
    }
}

File: src/main/java/equidiaryDB/EquidiaryDB.kt
package equidiaryDB

import com.fasterxml.jackson.databind.SerializationFeature
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import equidiaryDB.config.DBConfig
import equidiaryDB.config.NULL_CONFIG
import equidiaryDB.config.getDBConfig
import equidiaryDB.database.DataBaseE
import equidiaryDB.services.HorseService
import equidiaryDB.services.LoginService
import io.javalin.Javalin
import io.javalin.http.Context
import io.javalin.plugin.json.JavalinJackson
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger
import org.flywaydb.core.Flyway
import org.jetbrains.exposed.dao.id.IntIdTable
import org.jetbrains.exposed.sql.insert
import org.jetbrains.exposed.sql.javatime.date
import org.jetbrains.exposed.sql.transactions.transaction
import java.time.LocalDate

object Horses : IntIdTable("horse") {
    val name = varchar("horse_name", 150)
    val height = integer("height")
    val weight = integer("weight").nullable()
    val birthDate = date("birth_date")
}

object EquidiaryDB {
    private const val EQUIDIARYDB_PORT = 7001
    private lateinit var app: Javalin
    val equiLogger: Logger = LogManager.getLogger()
    lateinit var db: DataBaseE

    @JvmStatic
    fun start() {
        val config = getDBConfig()
        if (failedToLoadConfig(config)) {
            equiLogger.fatal("Failed to start EquidiaryDB")
            return
        }

        applyDataBaseMigrations(config)

        val objectMapper = jacksonObjectMapper()
        objectMapper.disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS)
        objectMapper.registerModule(JavaTimeModule())

        db = DataBaseE.createDatabase(config.host, config.port, config.user, config.password, config.schema, config.url)
        app = Javalin.create { it.jsonMapper(JavalinJackson(objectMapper)) }.start(EQUIDIARYDB_PORT)
        createEndPoints()

        transaction {
            Horses.insert {
                it[name] = "Fleur"
                it[height] = 160
                it[weight] = 500
                it[birthDate] = LocalDate.of(2015, 4, 22)
            }
        }
    }

    private fun applyDataBaseMigrations(config: DBConfig) {
        val flyway = Flyway
            .configure()
            .locations("classpath:migration")
            .dataSource(config.url, config.user, config.password)
            .load()

        flyway.migrate()
    }

    private fun failedToLoadConfig(config: DBConfig?) = config === NULL_CONFIG

    @JvmStatic
    fun stop() {
        app.stop()
    }

    private fun createEndPoints() {
        app.get("/") { ctx: Context -> ctx.result("Hello World") }
        app.get("/users/{name}") { ctx: Context ->
            ctx.result("Hello " + ctx.pathParam("name"))
        }
        app.post("/login") { LoginService() }
        app.get("/horse", HorseService())
    }
}
